lib/ui/painting/codec.cc

```diff
@@ -104,29 +110,123 @@ fxl::RefPtr<Codec> InitCodec(fml::WeakPtr<GrContext> context,
+    size_t trace_id) {
+  TRACE_FLOW_STEP("flutter", kInitCodecTraceTag, trace_id);
+  TRACE_EVENT0("blink", "InitCodecUncompressed");
+
+  if (buffer == nullptr || buffer->isEmpty()) {
+    FXL_LOG(ERROR) << "InitCodecUncompressed failed - buffer was empty";
+    return nullptr;
+  }
+
+  sk_sp<SkImage> skImage;
+  if (context) {
+    SkPixmap pixmap(image_info, buffer->data(), image_info.minRowBytes());
+    skImage = SkImage::MakeCrossContextFromPixmap(context.get(), pixmap, false,
+                                                  nullptr, true);
+  } else {
+    skImage = SkImage::MakeRasterData(image_info, std::move(buffer),
+                                      image_info.minRowBytes());
+  }
+
+  auto image = CanvasImage::Create();
+  image->set_image({skImage, unref_queue});
+  auto frameInfo = fxl::MakeRefCounted<FrameInfo>(std::move(image), 0);
+  return fxl::MakeRefCounted<SingleFrameCodec>(std::move(frameInfo));
+}
+
 void InitCodecAndInvokeCodecCallback(
     fxl::RefPtr<fxl::TaskRunner> ui_task_runner,
     fml::WeakPtr<GrContext> context,
     fxl::RefPtr<flow::SkiaUnrefQueue> unref_queue,
     std::unique_ptr<DartPersistentValue> callback,
     sk_sp<SkData> buffer,
+    std::unique_ptr<SkImageInfo> image_info,
     size_t trace_id) {
-  auto codec =
-      InitCodec(context, std::move(buffer), std::move(unref_queue), trace_id);
+  fxl::RefPtr<Codec> codec;
+  if (image_info) {
+    codec = InitCodecUncompressed(context, std::move(buffer), *image_info,
+                                  std::move(unref_queue), trace_id);
+  } else {
+    codec =
+        InitCodec(context, std::move(buffer), std::move(unref_queue), trace_id);
+  }
   ui_task_runner->PostTask(
       fxl::MakeCopyable([callback = std::move(callback),
                          codec = std::move(codec), trace_id]() mutable {
         InvokeCodecCallback(std::move(codec), std::move(callback), trace_id);
       }));
 }
 
+bool ConvertImageInfo(Dart_Handle image_info_handle,
+                      Dart_NativeArguments args,
+                      SkImageInfo* image_info) {
+  Dart_Handle width_handle = Dart_GetField(image_info_handle, ToDart("width"));
+  if (!Dart_IsInteger(width_handle)) {
+    Dart_SetReturnValue(args, ToDart("ImageInfo.width must be an integer"));
+    return false;
+  }
+  Dart_Handle height_handle =
+      Dart_GetField(image_info_handle, ToDart("height"));
+  if (!Dart_IsInteger(height_handle)) {
+    Dart_SetReturnValue(args, ToDart("ImageInfo.height must be an integer"));
+    return false;
+  }
+  Dart_Handle format_handle =
+      Dart_GetField(image_info_handle, ToDart("format"));
+  if (!Dart_IsInteger(format_handle)) {
+    Dart_SetReturnValue(args, ToDart("ImageInfo.format must be an integer"));
+    return false;
+  }
+
+  PixelFormat pixel_format = static_cast<PixelFormat>(
+      tonic::DartConverter<int>::FromDart(format_handle));
+  SkColorType color_type;
+  if (pixel_format == kRGBA8888) {
```

Comment:
Switch statement here with color_type set to unknown initially with a check that returns on the same after the switch. That way, adding to the enum and forgetting to handle that case will automatically give you a compile error.

Reply:
done